File: SHARE/REFERENCE/SUMMARY.R
library("tidyverse")

#READ
tb <- read_csv('data/tb.csv')


#PRINTING
print(df,n=1)
print(df %>% select(unemployment_rate))
print(df %>% pull(unemployment_rate) )



library('tidyverse')

print("----ORIGINAL----")
gap <- read_csv('data/gapminder.csv')
head(gap)

print("----GROUP BY CONTINENT----")
gap %>% group_by(continent)

print("----AVERAGE LIFE EXPECTANCY BY CONTINENT----")
gap %>% group_by(continent) %>% summarize(avgLifeExp = mean(lifeExp))


print("----AVERAGE LIFE EXPECTANCY BY COUNTRY----")
gap %>% group_by(country) %>% summarize(avgLifeExp = mean(lifeExp))

print("----AVERAGE LIFE EXPECTANCY BY CONTINENT AND YEAR----")
gap %>% group_by(continent, year) %>% summarize(mean(lifeExp))

print("----AVERAGE+MEDIAN EXPECTANCY BY YEAR----")
gap %>% group_by(year) %>% summarize(means = mean(lifeExp), medians = median(lifeExp))



# WE NEED TO MELT THIS DATASET TO MAKE IT TIDY
pew1 <- pew %>% pivot_longer(cols = c(-religion), names_to = 'income_groups', values_to = 'counts')
print(pew1)


#WE REVERSE THIS PROCESS BY PIVOTING THE DATA SET
pew2 <- pew1 %>% pivot_wider(id_cols = c(religion), names_from = 'income_groups', values_from = 'counts')
head(pew2,2)


#READ
tb <- read_csv('data/tb.csv')
print("ORIGINAL")
head(tb, 3)

#MELT
tb1 <- tb %>% pivot_longer(cols=m014:f65, names_to='demo', values_to='counts')
print("AFTER MELT")
head(tb1, 3)

#SPLIT
tb2 <- tb1 %>% separate(demo, c('gender','age'), sep = "[,]")
print("AFTER SPLIT ")
print(head(tb2, 3))



# TB <-  read_csv('data/tb1.csv') %>% 
#   pivot_longer(cols = starts_with('new'),names_to='demo', values_to='counts')
# head(TB, 3)

# TB1 <- TB %>% extract(demo, c('type','gender','age'),  regex = 'new[_]*([a-z]+)_([mf])([0-9]+$)')
# head(TB1)

#-----------------------------
print("----READ----")
#-----------------------------
mpg1 <- read_csv('data/mpg.csv')
head(mpg1)

#-----------------------------
print("----MUTATE----")
#-----------------------------
# OPTION-1: SEND MPG TO MUTATE
 mpg2 <- mutate(mpg1, cty = cty * 1.6/3.8,
         hwy = hwy * 1.6/3.8, 
         avg = (cty + hwy)/2)
head(mpg2); #print(mpg2['avg'])

#-----------------------------
print("----MUTATE----")
#-----------------------------
#OPTION-2: PIPE MPG INTO MUTATE 
mpg1 <- mpg1 %>% mutate(cty = cty * 1.6/3.8,
         hwy = hwy * 1.6/3.8, 
         avg = (cty + hwy)/2)
head(mpg1)

#-----------------------------
print("----SELECT----")
#-----------------------------
mpg2 <- mpg1 %>% select(year, manufacturer, avg)
head(mpg2)

#-----------------------------
print("----FILTER----")  
#-----------------------------
mpg2 <- mpg1 %>% filter(manufacturer=='audi')
head(mpg2)

#-----------------------------
print("----ARRANGE----")  
#-----------------------------
mpg2 <-mpg1 %>% arrange(-year)
head(mpg2)

#-----------------------------
print("----FILTER THEN ARRANGE----")  
#-----------------------------
#CHAIN TOGETHER 
mpg2 <-mpg1 %>% filter(manufacturer=='audi') %>% arrange(-year)  
head(mpg2,3)

#-----------------------------
#CONVERT TO FACTOR 
#-----------------------------
mpg2['model']=as.factor(mpg2$model)
head(mpg2,3)

#LOAD DATA
df1 <- read_csv('data/toy-1.csv')
df2 <- read_csv('data/toy-2.csv')


print("----ORIGINAL----")
head(df1)
head(df2)

print("----INNER JOIN----")
df1 %>% inner_join(df2) %>% head()
df1 %>% inner_join(df2, by = c('x')) %>% head()
df1 %>% inner_join(df2, by = c('x' , 'w')) %>% head()

print("----LEFT JOIN----")
df1 %>% left_join(df2) %>% head()
# df1 %>% left_join(df2,by = c('x' , 'y')) %>% head()

print("----RIGHT JOIN----")
df1 %>% right_join(df2) %>% head()

print("----FULL JOIN----")
df1 %>% full_join(df2) %>% head()


# library(tibble)
# tibble(mtcars)
# head(mtcars)
# glimpse(mtcars)
# glimpse(mtcars[3:4])
# colnames(mtcars)

File: SHARE/REFERENCE/SUMMARY.py

#-------------------------------
#NOTE THIS IS JUST OF LIST OF COMMANDS 
#FOR REFERENCE AND WILL NOT RUN
#-------------------------------

# IMPORT
import pandas as pd
 
# LOAD CSV
df = pd.read_csv('input.csv')

#SAVE CSV
df.to_csv('output.csv') #, sep='\t')
df.to_csv('output.csv',index=False) #dont write index 

#RENAME COLUMN
df=df.rename(columns={"year_C": "year"})

# MELTING
pd.melt(df1, id_vars =['ID']) 
pd.melt(df1, id_vars =['ID','H'])

# BY REPEATING FIRST 5 COLUMNS 0,1,2,3,4
df=pd.melt(df, id_vars =df.columns[0:5]); # print(df)

#REPLACE NAN 
dfA=dfA.fillna('False')

dfA["A"]=dfA["A"].fillna(False)  #one column only



# SPLITTING A LIST THAT IS STORED AS A STRING
# #Series.str()-- >Vectorized string functions for Series and Index.
df1 = pd.DataFrame([['[1, 2]','[3, 4]'], ['[NA, 6]','[7, 8]']], columns=['A', 'B'])
print("ORIGNIAL-\n",df1)
df1[['A1','A2']] = pd.DataFrame(df1.A.str.replace('[','', regex=True).str.replace(']','', regex=True).str.split(',').tolist())
df1=df1.drop(['A'],axis=1)                                 
print("MODIFIED-\n",df1)

#SERIES CONCATINATION
pd.concat([s1, s2],axis=0)
pd.concat([s1, s2],axis=1)

#DATAFRAME CONCATINATION
pd.concat([df1, df2],axis=0)
pd.concat([df1, df2],axis=1)

#JOINING
df1 = pd.DataFrame([['x1', 'y1'], ['x2', 'y2'], ['x3', 'y3']], columns=['x', 'y'])
df2 = pd.DataFrame([['x1', 'y1','w1', 'z1']
                  , ['x2','y2', 'w2', 'z2']
                  , ['x4','y4', 'w4', 'z4']],columns=['x', 'y','w', 'z'])
print("INPUTS:")
print(df1,'\n')
print(df2)
print("OUTPUTS:")
print('INNER JOIN: \n',df1.merge(df2, how='inner'))
print('LEFT JOIN:  \n',df1.merge(df2, how='left'))
print('RIGHT JOIN: \n',df1.merge(df2, how='right'))
print('OUTER JOIN: \n',df1.merge(df2, how='outer'))


#KEEP ONLY ROWS FOR WHICH COLUMN A='Y'
df = df[df['A'] == 'Y']

#REMOVE A COLUMN(S)
df = df.drop('A',axis=1)
df = df.drop(['A','B','C'],axis=1)

#MAKE A NEW COLUMN "A" BY SELECTING A SUB-STRING FROM col1
df["A"]  = df.col1.str[0:2]


#MAKE NEW ROW FROM BOOLEAN OPERATION ON ROW "B" 
df['A'] = df['B'].apply(lambda x: True if ((x=="cat") | (x=="dog")) else False)

# MUTATE 
mpg['cty'] = mpg['cty'] * 1.609/3.7854 # mile/gallon --> km/l
mpg['hwy'] = mpg['hwy'] * 1.609/3.7854
mpg['avg'] = (mpg.cty + mpg.hwy)/2

# SELECT
mpg[['year','manufacturer','avg']]

# FILTER 
mpg[mpg.manufacturer=='audi']

# ARRANGE 
 mpg.sort_values(by=['avg'], ascending=False)

# CHANGE DATA TYPE 
mpg['manufacturer']=mpg.manufacturer.astype('category')
mpg['model']=mpg.model.astype('string')


print("----GROUP BY CONTINENT----")
print((df.groupby('continent')).head())
# print('----\n',df.groupby('continent').tail())

print("----AVERAGE LIFE EXPECTANCY BY CONTINENT----")
print(df.groupby('continent')['lifeExp'].mean())

print("----AVERAGE LIFE EXPECTANCY BY COUNTRY----")
print(df.groupby('country')['lifeExp'].mean())

print("----AVERAGE LIFE EXPECTANCY BY CONTINENT AND YEAR----")
print(df.groupby(['continent','year'])['lifeExp'].mean())

print("----AVERAGE+MEDIAN EXPECTANCY BY YEAR----")
print(df.groupby('year')['lifeExp'].agg([np.mean,np.median]))

